### PR TITLE
Report current master in PeerFinder

### PR DESCRIPTION
**Pull request by DaveCTurner**
If we cannot join a cluster because the currently-elected master is
unreachable to us, but we can reach enough other master-eligible nodes,
then it can be a little hard to determine the problem from the logs: the
`ClusterFormationFailureHelper` reports that we have discovered a
possible quorum, but this is no help if we cannot contact the master
itself. The `PeerFinder` eventually starts to report the connection
failures, but it's hard to tell that they're the important onces.

This commit adjusts the logging from the `PeerFinder` to include an
indication that the connection we're attempting is to a node that was
recently believed to be the elected master by one of our peers.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/discovery/PeerFinderTests.java [hunk 1]
```diff
@@ -883,6 +883,48 @@ public boolean innerMatch(LogEvent event) {
...
+            .millis();
+
+        runAllRunnableTasks();
+
+        assertFoundPeers(otherNode);
+        final DiscoveryNode unreachableMaster = newDiscoveryNode("unreachable-master");
+        transportAddressConnector.unreachableAddresses.add(unreachableMaster.getAddress());
+
+        MockLogAppender appender = new MockLogAppender();
+        try (var ignored = appender.capturing(PeerFinder.class)) {
```
Comment (idegtiarenko) [inline server/src/test/java/org/elasticsearch/discovery/PeerFinderTests.java:903]: NIT: consider using `assertThatLogger` to minimize the setup

**Comment by DaveCTurner** [comment]
`CoordinationDiagnosticsServiceTests` still needs work, bear with me here.

File: server/src/main/java/org/elasticsearch/discovery/PeerFinder.java [hunk 2]
```diff
@@ -352,10 +352,17 @@ protected void startProbe(TransportAddress transportAddress) {
         }
     }
 
+    public List<DiscoveryNode> getMastersOfPeers() {
+        synchronized (mutex) {
+            return peersByAddress.values().stream().flatMap(p -> p.lastKnownMasterNode.stream()).distinct().toList();
+        }
+    }
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/discovery/PeerFinder.java:359]: I believe this is used to populate the new `List<DiscoveryNode> mastersOfPeers`. I was originally assuming we report master seen by each node to verify each have the same, however here it is using `distinct()`. Why not using a set to make it obvious there are no duplicates?

**Review by idegtiarenko — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/discovery/PeerFinder.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/discovery/PeerFinder.java:360]: Fair point, done in 8802ebf

**Pull request merged**